FCBHDBP-359

vichugofsl opened the pull request:
# Description
It has changed the method `generateCacheString` to create a well formed key to store the values into `memcached`. It has used `base64_encode` method to make sure all characters are valid. Also, If length of value is more than 240 we will create a key using md5 and sha1 algorithms and we will try to avoid hash collisions using both hash algorithms and additionally using the key length.

## Issue Link
https://fullstacklabs.atlassian.net/browse/FCBHDBP-359

## How Do I QA 
- Language search url encode
    https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-6b3d5bf2-770b-44f1-bc4a-f53083921a40

- Bibles search without url encode characters
    https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-0c02737c-57f9-4ac9-9797-af368e25df22

- Bibles search url encode
    https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-52fb2655-2af4-4a95-9614-a1f0a8b9eb62

- Bibles search url encode mixed
    https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-62c0050f-d17b-49db-810d-90ab4d1c5eb5

- Countries search url encode
    https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-0baca334-40ae-4b1a-9f3c-5cab14550dec

vichugofsl commented:
@bradflood , could you check if this PR will be ok to handle the ticket [FCBHDBP-359](https://fullstacklabs.atlassian.net/browse/FCBHDBP-359)?

bradflood commented:
@vichugofsl, two points:
1)  independent of how we encode the cache key, we still need to url-decode the user-provided search string;
2) Rather than changing the cache key to a hash for every key, let's instead just use the hash algorithm you describe for the search keys. This limits the scope of this ticket.

Also, there are a couple comments about changes to the lock acquisition:
1) change timeout from 45 seconds to 20 seconds
2) change key length check from 240 characters to 230 characters

bradflood reviewed (changes requested):
@vichugofsl see comments elsewhere regarding changes

vichugofsl commented on app/Http/Controllers/Bible/BiblesController.php:
```diff
@@ -269,10 +269,10 @@ public function search($search_text)
-        );
-        $bibles = cacheRemember('bibles_search', $cache_params, now()->addDay(), function () use ($key, $limit, $formatted_search) {
+        $cache_params = [$limit, $page, $formatted_search_cache, $key];
+        $cache_key = generateCacheSafeKey('countries', $cache_params);
```
We will create a safe key before we call the cacheRememberByKey method. For now, it will be implemented for the searching endpoints

vichugofsl commented on app/Http/Controllers/Wiki/LanguagesController.php:
```diff
@@ -212,10 +211,11 @@ public function search($search_text)
-            'languages_search',
-            $cache_params,
+        ;
+        $cache_key = generateCacheSafeKey('languages_search', $cache_params);
```
We will create a safe key before we call the cacheRememberByKey method. For now, it will be implemented for the searching endpoints

vichugofsl commented on app/Http/Controllers/Wiki/CountriesController.php:
```diff
@@ -128,10 +128,10 @@ public function search($search_text = '')
-        );
-        $countries = cacheRemember('countries', $cache_params, now()->addDay(), function () use ($limit, $page, $formatted_search) {
+        $cache_params = [$GLOBALS['i18n_iso'], $limit, $page, $formatted_search_cache];
+        $cache_key = generateCacheSafeKey('countries', $cache_params);
```
We will create a safe key before we call the cacheRememberByKey method. For now, it will be implemented for the searching endpoints

vichugofsl commented on app/Helpers/Helpers.php:
```diff
@@ -123,13 +125,35 @@ function createCacheLock($cache_key, $lock_timeout = 10)
+ * @param Closure $callback
+ * @return mixed
+ */
+function cacheRememberByKey(string $key, Carbon $ttl, Closure $callback)
```
This new method was created to store the value into cache but the key will be created before. So, we can call the method `generateCacheSafeKey` to create a safe key.

vichugofsl commented on app/Helpers/Helpers.php:
```diff
@@ -169,29 +194,79 @@ function cacheRemember($cache_key, $cache_args, $ttl, $callback)
+ */
+function transformQuerySearchText(string $search_text): string
+{
+    $formatted_search = urldecode($search_text);
```
We will call the urldecode method before everything.

vichugofsl commented:
> @vichugofsl, two points:
> 
>     1. independent of how we encode the cache key, we still need to url-decode the user-provided search string;
> 
>     2. Rather than changing the cache key to a hash for every key, let's instead just use the hash algorithm you describe for the search keys. This limits the scope of this ticket.
> 
> 
> Also, there are a couple comments about changes to the lock acquisition:
> 
>     1. change timeout from 45 seconds to 20 seconds
>        function cacheRemember($cache_key, $cache_args, $ttl, $callback)
>        {
>        $key = generateCacheString($cache_key, $cache_args);
>        **// if something fails on the callback, release the lock
>        // 20 seconds was selected to allow for the longest query to complete.
>        // This is not based on any empirical evidence.
>        $lock_timeout = 20;**
>        $value = Cache::get($key);
> 
> 
>     2. when logging exception, add key
>        // cache not set. try to acquire lock to gain access to the callback
>        $lock = createCacheLock($key);
>        if ($lock->acquire()) {
>        try {
>        // lock acquired. access resource via callback
>        $value = $callback();
>        if (!is_null($value)) {
>        Cache::put($key, $value, $ttl);
>        } else {
>        Log::error("CacheRemember. callback returned null for key: " . $key);
>        }
>        $lock->release();
>        return $value;
>        } catch (Exception $exception) {
>        $lock->release();
>        **Log::error("Exception trying to acquire lock for key ["+$key+"]" . $exception);**
>        throw $exception;
>        }
> 
>     3. change key length check from 240 characters to 230 characters, and adjust the comment also

@bradflood
- we are call url-decode method to decode the user-provided search string. It is used into the `transformQuerySearchText` method.
- The new way to generate a safe key will only be used for the searching endpoints. The above is related with the new methods: `generateCacheSafeKey` and `cacheRememberByKey`. 
- The $lock_timeout has been updated to 20-

bradflood commented on app/Http/Controllers/APIController.php:
```diff
@@ -376,10 +376,7 @@ protected function getKey()
-        $formatted_search = preg_replace('/[+\-><\(\)~*\"@%]+/', ' ', $formatted_search);
-        $formatted_search = preg_replace('!\s+!', ' ', $formatted_search);
-        return trim($formatted_search);
+        return transformQuerySearchText($search_text);
```
@vichugofsl I may be misunderstanding the syntax, but this seems like it will result in a tight loop. Please explain

vichugofsl replied:
@bradflood , I understand you. But, it won't be a tight loop because it is calling to the method that we have defined in the helper: `app/Helpers/Helper.php`. The case is that they (helper method and class method) have the same name.

It will be a tight loop if we would call the class method e.g.

```php
public function transformQuerySearchText(string $search_text): string
{
    return $this->transformQuerySearchText($search_text);
}
```

bradflood replied:
thank you for the explanation

vichugofsl commented:
> 2\. Exception

@bradflood , I have updated the comment and I have added the key in the exception messages.

e.g.

```php
if (is_null($value)) {
    // !!! **** my assumption about when the cache value will be available is not valid
    throw new Exception("Exception when the cache value is null for key [".$key."]");
}
```

```php
throw new \UnexpectedValueException("Undefined Error for key [".$key."]");
```
